File: python/1930. Unique Length-3 Palindromic Subsequences.py
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        first_occ = {}
        last_occ = {}
        for i, ch in enumerate(s):
            if not ch in first_occ:
                first_occ[ch] = i
            last_occ[ch] = i
        result = 0

        for ch in first_occ:
            first = first_occ[ch]
            last = last_occ[ch]
            if first < last:
                uniq_sym = set(s[first+1:last])
                result += len(uniq_sym)
        return result

File: rust/1930. Unique Length-3 Palindromic Subsequences.rs
use std::collections::HashMap;
use std::collections::HashSet;

impl Solution {
    pub fn count_palindromic_subsequence(s: String) -> i32 {
        let mut first_occ: HashMap<char, usize> = HashMap::new();
        let mut last_occ: HashMap<char, usize> = HashMap::new();
        for (i, ch) in s.chars().enumerate() {
            first_occ.entry(ch).or_insert(i);
            last_occ.insert(ch, i);
        }

        let mut result = 0;

        for (ch, first) in first_occ.iter() {
            let last = last_occ.get(ch).unwrap();
            if first < last {
                let unique_sym = s[first + 1..*last].chars().collect::<HashSet<_>>();
                result += unique_sym.len();
            }
        }

        result as i32
    }
}
